dba/index_and_backup.go

```diff
@@ -0,0 +1,96 @@
+package dba
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"time"
+
+	"github.com/blevesearch/bleve"
+	"github.com/blevesearch/bleve/document"
+)
+
+const (
+	// BackupFolderLocation is the folder where backup json dumps go to.
+	BackupFolderLocation = "/tmp"
+
+	// UUIDsIndexName is the name of index that is used to store UUIDs of the
+	// resources.
+	UUIDsIndexName = "UUIDsIndex"
+)
+
+// indexAndBackup wraps bleve.Index Document and Index methods. In addition it
+// triggers backing up resources data to disk on Index operation.
+type indexAndBackup struct {
+	index bleve.Index
+}
+
+func (i *indexAndBackup) Index(id string, data interface{}) error {
+	if err := i.index.Index(id, data); err != nil {
+		return err
+	}
+
+	go i.backup()
+
+	return nil
+}
+
+func (i *indexAndBackup) Document(id string) (*document.Document, error) {
+	return i.index.Document(id)
+}
+
+func (i *indexAndBackup) backup() {
+	uuids, err := i.Document(UUIDsIndexName)
+	if err != nil {
+		log.Println("indexAndBackup#backup: ERR getting index: ", UUIDsIndexName, err)
+		return
+	}
+
+	log.Println("indexAndBackup#backup: About to backup: ", len(uuids.Fields), "from index: ", UUIDsIndexName)
+
+	listOfResources := []map[string]interface{}{}
+
+	for _, uuid := range uuids.Fields {
+		resource, err := i.Document(string(uuid.Value()))
+		if err != nil {
+			log.Println("indexAndBackup#backup: ERR finding resource: ", string(uuid.Value()), err)
+			return
+		}
+
+		dd, err := MakeMapFromDocument(resource)
```

Comment:
Peter: "dd, what does that stand for? Dare Devil?" 

Senthil's response: "I didn't want to name it just 'd', so I called it 'dd'". 

FYI to future people: dd is a meta syntactic variable (like "foo" or "bar). But in this case, the fact that it begins with the fourth letter in the English language ('d') adds a bit of spice and should give the code reader the (correct) suspicion that this related to "document", and we might infer that the second 'd' is short for 'dictionary', which is actually a **d**ecent fit in this case! 

So in short, it's all good.